### PR TITLE
Huggingface speech recognition finetuning examples

### PR DESCRIPTION
**Pull request by footprinthere**

### CODE REVIEW

File: examples/huggingface/speech_recognition/ctc/script.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Move data directories
+mv ../data/downloads ../downloads
+mv ../data/common_voice ../common_voice
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech_recognition/ctc/script.sh:3]: Remove these lines, and set `dataset_cache_dir` to `/workspace/data`

File: examples/huggingface/speech_recognition/ctc/script.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+echo "\n### INSTALLING GCC / G++"
+apt-get install gcc -qq
+apt-get install g++ -qq
+
+echo "\n### INSTALLING LIBSNDFILE"
+apt-get install libsndfile-dev -qq
+
+echo "\n### INSTALLING REQUIREMENTS"
+pip install torch==1.10.0+cu113 torchvision==0.11.1+cu113 torchaudio==0.10.0+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html
+pip install git+https://github.com/huggingface/transformers
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech_recognition/ctc/script.sh:17]: I think we need to specify the transformers version too, if we use specific torch version.

File: examples/huggingface/speech_recognition/seq2seq/script.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+apt-get install gcc -qq
+apt-get install g++ -qq
+
+echo "\n### INSTALLING LIBSNDFILE"
+apt-get install libsndfile-dev -qq
+apt-get install libsndfile1 -qq
+
+echo "\n### INSTALLING PIP REQUIREMENTS"
+pip install torch==1.10.0+cu113 torchvision==0.11.1+cu113 torchaudio==0.10.0+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html
+pip install git+https://github.com/huggingface/transformers
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech_recognition/seq2seq/script.sh:12]: Same comment

**Comment by Hyeonmin-Ha** [comment]
We need to update the directory structure.
Currently, language-related HF examples are in `examples/huggingface`.
But, speech-related HF examples in `examples/hug../speech_recognition`.
One way to resolve this will be move lang-related HF examples to `examples/hug../language`.

File: examples/huggingface/speech_recognition/seq2seq/run_speech_recognition_seq2seq.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+        # cut bos token here as it's append later anyways
+        if (labels[:, 0] == self.decoder_start_token_id).all().cpu().item():
+            labels = labels[:, 1:]
+
+        batch["labels"] = labels
+
+        return batch
+
+
+def prepare_model(model_path):
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech_recognition/seq2seq/run_speech_recognition_seq2seq.py:239]: How about move this function to a separated script (`prepare.py` for example) for consistency with the original example?

**Review by Hyeonmin-Ha — CHANGES_REQUESTED** [review]

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 2 above)
Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh:17]: And move this to `requirements.txt`

Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh:17]: Do we need torch 1.10? `requirements.txt` says torch>=1.5.

**Review by Hyeonmin-Ha — CHANGES_REQUESTED** [review]

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 2 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/ctc/script.sh:17]: 1. Specifying any version of `transformers` results to an ImportError:
    ```
    ImportError: : This example requires a source install from HuggingFace Transformers (see `https://huggingface.co/transformers/installation.html#installing-from-source`), but the version found is 4.21.1.
    ```
    As the [docs](https://huggingface.co/docs/transformers/installation#install-from-source) says, it seems that it is the only way to install `transformers` directory from git source.

2. Moved to `requirements.txt`.

3. Similar to the problem that happened in `MMF` example. Without explicitly installing `torch`, the program would result to an NCCL error and fail.

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 1 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/ctc/script.sh]: That results to an error in this example.
```
formats: can't open input file `downloads/extracted/14fbc5e0dd3e55b1771c258791de7514a56105ebe0f58e6aec60cae718283064/cv-corpus-6.1-2020-12-11/hi/clips/common_voice_hi_23795238.mp3': No such file or directory
```
It seems that regardless of the path specified by `dataset_cache_dir`, the program always tries to find datasets in the directory where the program code is placed. I have tried to find out why but failed.

File: examples/huggingface/speech_recognition/seq2seq/run_speech_recognition_seq2seq.py (hunk 4 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/seq2seq/run_speech_recognition_seq2seq.py:239]: Comment reflected.

File: examples/huggingface/speech_recognition/seq2seq/script.sh (hunk 3 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/seq2seq/script.sh:12]: Please see the comment above.

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 2 above)
Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh]: Have you tried `pip install git+https://github.com/django/django.git@<tag name>` where the tag name specifies the release you are using?

File: examples/huggingface/speech/speech_recognition_ctc/script.sh [hunk 5]
```diff
@@ -1,10 +1,12 @@
 # Move data directories
 mv ../data/downloads ../downloads
 mv ../data/common_voice ../common_voice
+rm -rf data
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech/speech_recognition_ctc/script.sh:4]: What does this actually remove?
This is dependent with the docker image we use, which specifies the start point.

File: examples/huggingface/speech/speech_recognition_seq2seq/model.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+from transformers import (
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech/speech_recognition_seq2seq/model.py:1]: Please rename this script to `prepare_model.py`.

File: examples/huggingface/speech/speech_recognition_seq2seq/run_speech_recognition_seq2seq.py [hunk 7]
```diff
@@ -53,6 +51,8 @@
 
 import periflow_sdk as pf
 
+from model import prepare_model
+
```
Comment (Hyeonmin-Ha) [inline examples/huggingface/speech/speech_recognition_seq2seq/run_speech_recognition_seq2seq.py:55]: Please run `prepare_model` separately with this script.
i.e, Run `prepare_model.py`, and then run this script in `script.sh`. (As in the original example)
The reason for this is that we'd like to keep the original code.

**Review by Hyeonmin-Ha — CHANGES_REQUESTED** [review]

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 1 above)
Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh]: It seems that [eval dataset loading](https://github.com/friendliai/periflow-python-sdk/blob/example-speech/examples/huggingface/speech/speech_recognition_ctc/run_speech_recognition_ctc.py#L457) doesn't use `cache_dir`.
Please try fixing this.

**Comment by Hyeonmin-Ha** [comment]
Please resolve the conflict.

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 2 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/ctc/script.sh]: The [programs](https://github.com/friendliai/periflow-python-sdk/blob/70db9c177f85cc0a1158216b781b7a0bbcf8cfd3/examples/huggingface/speech/speech_recognition_ctc/run_speech_recognition_ctc.py#L58) says we need version `4.22.0.dev0`, but it seems that the `transformers` [github repository](https://github.com/huggingface/transformers/tags) does not support this version as a release or a tag.

File: examples/huggingface/speech/speech_recognition_seq2seq/run_speech_recognition_seq2seq.py (hunk 7 above)
Comment (footprinthere) [reply examples/huggingface/speech/speech_recognition_seq2seq/run_speech_recognition_seq2seq.py:55]: Comment reflected.

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 1 above)
Comment (footprinthere) [reply examples/huggingface/speech_recognition/ctc/script.sh]: Or you can check the job log of ___1967___.

File: examples/huggingface/speech_recognition/ctc/script.sh (hunk 2 above)
Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh]: I think we are using too recent example code, which is before the release.
The latest transformers version is 4.21, so we cannot meet the requirement with the stable version.
Let's skip this issue and resolve later.

Comment (Hyeonmin-Ha) [reply examples/huggingface/speech_recognition/ctc/script.sh]: And as you said in the other PR, our docker image uses CUDA 11.6, so we should use torch that support the CUDA version. (i.e, torch 1.12)

**Comment by footprinthere** [comment]
All problems solved. @Hyeonmin-Ha

**Review by Hyeonmin-Ha — APPROVED** [review]

**Pull request merged**